### PR TITLE
gossip: make sleeps react to shutdown signal

### LINKED ISSUE #3749: gossip: Make the sleep code react to stop
There are multiple usage of sleep in the gossip code base. The sleep lasts for 30s by default which will hang the stop operation.

### PR DESCRIPTION
**Pull request by piodul**
This change makes most sleeps in gossip.cc abortable. It is now possible to quickly shut down a node during startup, most notably during the phase while it waits for gossip to settle.

Fixes #3749.

Tests: unit(dev)

### CODE REVIEW

File: gms/gossiper.hh [hunk 1]
```diff
@@ -548,7 +548,6 @@ public:
     future<> add_local_application_state(std::initializer_list<std::pair<application_state, utils::in<versioned_value>>>);
 
     // Needed by seastar::sharded
-    future<> stop();
```
Comment (tgrabiec) [inline gms/gossiper.hh]: You should also remove `// Needed by seastar::sharded`, which applied top `stop()` only.

File: gms/gossiper.cc [hunk 2]
```diff
@@ -125,10 +125,11 @@ class feature_enabler : public i_endpoint_state_change_subscriber {
     void on_restart(inet_address, endpoint_state) override {}
 };
 
-gossiper::gossiper(feature_service& features, db::config& cfg)
-        : _feature_service(features)
+gossiper::gossiper(feature_service& features, db::config& cfg, abort_source& as)
```
Comment (tgrabiec) [inline gms/gossiper.cc]: I'm not sure interrupting all services at once is the right approach in general.

Shouldn't we rather have the stop operation (`gossiper::stop_gossiping()`) trigger the interruption of internal operations, on a private abort_source? So that we don't interrupt anything until it's really time to stop the component. Just theorizing now, there could be another component which is relying on the gossiper to complete its (the component's) operations, so we should first let that component to stop itself (which may want to wait for those operations to complete), and only then stop the gossiper.

Comment (piodul) [reply gms/gossiper.cc]: We already use a similar approach in `storage_proxy.cc` - take global `abort_source`, and use `sleep_abortable` with it everywhere.

I can see that all paths that pass through `sleep_abortable`s from gossiper.cc are called either from HTTP API (directly or indirectly), or from initialization. When initialization is aborted, it is desirable to stop waiting. As for HTTP API, it will just fail the request if user is unlucky enough to perform it during shutdown.

File: gms/gossiper.hh (hunk 1 above)
Comment (piodul) [reply gms/gossiper.hh]: Removed the comment as well

File: gms/gossiper.cc (hunk 2 above)
Comment (tgrabiec) [reply gms/gossiper.cc]: Do you mean `storage_service`? I don't see abort_source usage in storage_proxy.cc. It's true that storage_service uses this approach, I still think it's a bad approach in general.

It's also true that this approach is not currently harmful given what operations it affects.

Comment (tgrabiec) [reply gms/gossiper.cc]: Another argument for the alternative approach is that we want to stop the gossiper operations not only on ^C, but also when exception interrupts the boot process. Making `stop()` trigger abort handles both cases.

Comment (piodul) [reply gms/gossiper.cc]: Sorry, I meant `storage_service.cc`, my mistake.

I agree with you argument, I remade aborting so as to abort on a private `abort_source`. I left aborting on Ctrl+C `abort_source` in initialization code.

File: gms/gossiper.cc [hunk 3]
```diff
@@ -1739,15 +1739,16 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
         }).then([this, generation] {
             logger.trace("gossip started with generation {}", generation);
             _enabled = true;
+            _abort_source = abort_source();
```
Comment (tgrabiec) [inline gms/gossiper.cc]: Why is this line needed? _abort_source should be already constructed in the constructor by then.

File: gms/gossiper.cc [hunk 4]
```diff
@@ -1133,7 +1133,7 @@ future<> gossiper::assassinate_endpoint(sstring address) {
                 auto ring_delay = std::chrono::milliseconds(gossiper._cfg.ring_delay_ms());
                 logger.info("Sleeping for {} ms to ensure {} does not change", ring_delay.count(), endpoint);
                 // make sure it did not change
-                sleep(ring_delay).get();
+                sleep_abortable(ring_delay, gossiper._abort_source).get();
```
Comment (tgrabiec) [inline gms/gossiper.cc]: This sleep is part of the `assassinate_endpoint()` operation (the operation joins with this, it doesn't run in the background). The sleeps in `wait_for_gossip_to_settle()` are also like that. In the latter case we chose the caller to interrupt the operation (by providing an abort_source as an argument). In the former case, we chose stop_gossiping() to interrupt it. We could have let the caller interrupt `assasinate_endpoint()` as well. I wonder why did you chose otherwise.

The idea was that whoever uses the gossiper (having an active foreground operation counts as a use), should be stopped before the gossiper is stopped. And when the gossiper is stopped, it interrupts only operations started by itself, which don't count as external use.

Comment (piodul) [reply gms/gossiper.cc]: I converted other methods to take an abort source as well. The only internal task I see is `gossiper::run` and it does not sleep, so we may drop the `abort_source` altogether.

I observed that we do not wait for HTTP requests to finish before stopping the gossiper, so it is possible for somebody to start, for example, an `assasinate` operation while we are shutting down. I'm not sure what to do about it - guard such operations with a gate?

Comment (tgrabiec) [reply gms/gossiper.cc]: Before the gossiper is stopped (sharded::stop), the API server should be stopped. Or at least the part of the API which uses the gossiper. Stopping is interrupt (abort_source) + await (gate for tracking).

File: gms/gossiper.cc [hunk 5]
```diff
@@ -1099,18 +1099,18 @@ future<> gossiper::advertise_token_removed(inet_address endpoint, utils::UUID ho
...
-    return assassinate_endpoint(address);
+    return assassinate_endpoint(address, as);
 }
 
-future<> gossiper::assassinate_endpoint(sstring address) {
-    return get_gossiper().invoke_on(0, [address] (auto&& gossiper) {
-        return seastar::async([&gossiper, g = gossiper.shared_from_this(), address] {
+future<> gossiper::assassinate_endpoint(sstring address, abort_source& as) {
+    return get_gossiper().invoke_on(0, [address, &as] (auto&& gossiper) {
+        return seastar::async([&gossiper, g = gossiper.shared_from_this(), address, &as] {
```
Comment (tgrabiec) [inline gms/gossiper.cc]: `abort_source` is not safe to use across shards (across threads). It's a shard-local object.

invoke_on() may execute the lambda on a different shard if the current shard is not 0.

File: api/gossiper.cc [hunk 6]
```diff
@@ -57,12 +57,13 @@ void set_gossiper(http_context& ctx, routes& r) {
     });
 
     httpd::gossiper_json::assassinate_endpoint.set(r, [](std::unique_ptr<request> req) {
+        abort_source as;
```
Comment (nyh) [inline api/gossiper.cc]: This variable is on the stack, below you pass it by reference to asynchronous functions which keep it after this function returns a future. I think you need to use some existing abort source, or do_with()?

File: gms/gossiper.cc (hunk 4 above)
Comment (piodul) [reply gms/gossiper.cc]: Proper shutdown of services seems to be a large topic (https://github.com/scylladb/scylla/issues/293). I think this is a topic too big to be handled here. I tried calling `sharded::stop` on the gossip service, but some other services still tried to use it during shutdown, which caused the nodes on my local cluster to crash.

I don't see a way to unregister a subset of HTTP routes, so I added a gate for gossiper methods that face the HTTP API. The gate is closed during shutdown operation (via `defer_verbose_shutdown`). Those routes will just return an error in case they are called during shutdown.

File: gms/gossiper.cc (hunk 5 above)
Comment (piodul) [reply gms/gossiper.cc]: I removed the `abort_source&` argument from the function. It wouldn't be used anyway (as we don't handle cancellation of HTTP requests), and I can't think of a solution for passing the abort source that would look nice.

**Comment by piodul** [comment]
Additional change: I removed the commit that gets rid of `gossiper::stop` function. Although this function is not necessary, it looks like it is a convention for services, so I would like to keep it.

File: gms/gossiper.cc [hunk 7]
```diff
@@ -1064,45 +1064,49 @@ future<> gossiper::replicate(inet_address ep, application_state key, const versi
...
-        auto ring_delay = std::chrono::milliseconds(_cfg.ring_delay_ms());
-        logger.info("Sleeping for {}ms to ensure {} does not change", ring_delay.count(), endpoint);
-        sleep(ring_delay).get();
-        // make sure it did not change
-        auto& eps = get_endpoint_state(endpoint);
-        if (eps.get_heart_beat_state().get_generation() != generation) {
-            throw std::runtime_error(format("Endpoint {} generation changed while trying to remove it", endpoint));
-        }
+future<> gossiper::advertise_removing(inet_address endpoint, utils::UUID host_id, utils::UUID local_host_id, abort_source& as) {
+    return seastar::with_gate(_gate, [this, g = this->shared_from_this(), endpoint, host_id, local_host_id, &as] {
```
Comment (asias) [inline gms/gossiper.cc]: You can use a half indentation to reduce the changes and restore the indentation later.

File: gms/gossiper.cc [hunk 8]
```diff
@@ -1064,45 +1064,49 @@ future<> gossiper::replicate(inet_address ep, application_state key, const versi
...
-        auto expire_time = compute_expire_time();
-        eps.add_application_state(application_state::STATUS, _value_factory.removed_nonlocal(host_id, expire_time.time_since_epoch().count()));
-        logger.info("Completing removal of {}", endpoint);
-        add_expire_time_for_endpoint(endpoint, expire_time);
-        endpoint_state_map[endpoint] = eps;
-        replicate(endpoint, eps).get();
-        // ensure at least one gossip round occurs before returning
-        sleep(INTERVAL * 2).get();
+future<> gossiper::advertise_token_removed(inet_address endpoint, utils::UUID host_id, abort_source& as) {
+    return seastar::with_gate(_gate, [this, g = this->shared_from_this(), endpoint, host_id, &as] {
```
Comment (asias) [inline gms/gossiper.cc]: Same here.

File: gms/gossiper.cc [hunk 9]
```diff
@@ -1112,54 +1116,56 @@ future<> gossiper::unsafe_assassinate_endpoint(sstring address) {
...
+                        throw std::runtime_error(format("Unable to calculate tokens for {}", endpoint));
+                    }
 
-                int generation = ep_state.get_heart_beat_state().get_generation();
-                int heartbeat = ep_state.get_heart_beat_state().get_heart_beat_version();
-                auto ring_delay = std::chrono::milliseconds(gossiper._cfg.ring_delay_ms());
-                logger.info("Sleeping for {} ms to ensure {} does not change", ring_delay.count(), endpoint);
-                // make sure it did not change
-                sleep(ring_delay).get();
+                    int generation = ep_state.get_heart_beat_state().get_generation();
```
Comment (asias) [inline gms/gossiper.cc]: Extra white space?

**Comment by piodul** [comment]
@asias I used half-indentation as you suggested, and fixed whitespace in a separate commit.

File: main.cc [hunk 10]
```diff
@@ -962,7 +962,7 @@ int main(int ac, char** av) {
             //FIXME: discarded future
             (void)api::set_server_cache(ctx);
             startlog.info("Waiting for gossip to settle before accepting client requests...");
-            gms::get_local_gossiper().wait_for_gossip_to_settle().get();
+            gms::get_local_gossiper().wait_for_gossip_to_settle(stop_signal.as_local_abort_source()).get();
```
Comment (asias) [inline main.cc]: In  storage_service class, we have 
 
   abort_source& _abort_source

member. Why not do this to gossiper class as well so we can avoid passing abort_source to all the calls.

File: gms/gossiper.cc [hunk 11]
```diff
@@ -1065,7 +1065,8 @@ future<> gossiper::replicate(inet_address ep, application_state key, const versi
 }
 
 future<> gossiper::advertise_removing(inet_address endpoint, utils::UUID host_id, utils::UUID local_host_id, abort_source& as) {
-    return seastar::async([this, g = this->shared_from_this(), endpoint, host_id, local_host_id, &as] {
+  return seastar::with_gate(_gate, [this, g = this->shared_from_this(), endpoint, host_id, local_host_id, &as] {
```
Comment (asias) [inline gms/gossiper.cc]: class gossiper inherits from class seastar::async_sharded_service. This guarantees sharded::stop() will wait until all the references disappear before returning. This looks redundant to me. In general I like using gate to wait for the existing and prevent further operations, but it is not enough just for those functions you changed here. 

So, I would recommend to contain only the abort source changes in this PR so we can get it merged faster.

**Comment by asias** [comment]
> @asias I used half-indentation as you suggested, and fixed whitespace in a separate commit.

Thanks. It is much easier to review now!

File: main.cc (hunk 10 above)
Comment (piodul) [reply main.cc]: The idea is to let the caller of the operation supply an `abort_source`, so that they control when the operation is cancelled. Initially, I did it in the way you described, but changed after comments from @tgrabiec.

File: gms/gossiper.cc (hunk 11 above)
Comment (piodul) [reply gms/gossiper.cc]: OK, I dropped the last two commits (related to the gate).

File: main.cc (hunk 10 above)
Comment (asias) [reply main.cc]: There is only one caller that is storage_service.cc. So every abort source passed to gossiper is from storage_service which in turn from the object from main.cc. What's the point to make storage_service to pass abort source?

**Comment by slivne** [comment]
@asias ping - please approve if its ok now

**Comment by asias** [comment]
> @asias ping - please approve if its ok now

We have pending discussion on passing the abort source to gossip object instead of passing to gossip member functions everywhere. 

@piodul I do not see practical benefits to passing abort source to gossip member functions everywhere.

**Comment by piodul** [comment]
@asias I changed the PR so that the gossiper holds a reference to the abort source in a field. All but one sleeps are now abortable. I left one non-abortable sleep in `do_stop_gossiping` because it is on a path that may run at shutdown and it seemed intentional.

**Comment by asias** [comment]
Thanks @piodul . It looks good to me.

**Comment by slivne** [comment]
@tgrabiec / @nyh - comments seems to been have handled - @asias blessed this - please re-review

**Comment by nyh** [comment]
Looks good to me, but there is one open discussion with @tgrabiec about details of the shutdown order that I didn't understand enough to be able to judge whether this discussion was resolved or not. Tomek?

File: main.cc (hunk 10 above)
Comment (tgrabiec) [reply main.cc]: @asias In general, having one abort source per component is inflexible.

First, we may want to interrupt operations not only when the user hits ^C. If an operation on gossiper is part of a background operation started by storage_service, we will want to interrupt that operation when storage_service::stop() is called. When we fail to boot with an exception we will trigger a cascade of stops on each component (according to the depenency chain), without the main `abort_source` being triggered. 

Another problem is that, in general, interrupting all operations from all components at once may not be correct due to dependencies. If component A depends on B, then A needs to be stopped before B is stopped. If you interrupt B before stopping A, then A may not be able to stop gracefully. B::stop() should interrupt background operations managed by B, only after A stopped itself.

Those are theoretical problems though for now, so I committed the patch.
